tea.py

```diff
@@ -0,0 +1,443 @@
+
+        # Concatenate the dataframes side by side
+        df = pd.concat([components, volumes, masses, densities], axis=1)
+        df.set_index('Components', inplace=True)
+        df.index.name = None
+        print(df)
+        return df
+    
+    def print_stack_energy_densities(self, print_dict=False):
+        if not hasattr(self, "stack_edd"):
+                self.calculate_stack_energy_densities()
+        if print_dict == True:
+            print(dict)
+            
+    def calculate_stack_energy_densities(self):
+        # defaults for current collector thicknesses and density, as well as for electrolyte density are provided
+        self.stack_edd = {} # stack energy densities dict
+
+        # ocv's
+        ### add optional user defined cut-off voltages (supplied with inputs)?
+        ne_ocv = self.parameter_values["Negative electrode OCP [V]"]
+        pe_ocv = self.parameter_values["Positive electrode OCP [V]"]
+        x0, x100, y100, y0 = pybamm.lithium_ion.get_min_max_stoichiometries(
+            self.parameter_values
+        )
+        soc = pybamm.linspace(0, 1)
+        x = x0 + soc * (x100 - x0)
+        y = y0 - soc * (y0 - y100)
+
+        self.stack_edd["Negative electrode average OCP [V]"] = self.parameter_values.get("Negative electrode average OCP [V]") or ne_ocv(x).entries.mean()
+        self.stack_edd["Positive electrode average OCP [V]"] = self.parameter_values.get("Positive electrode average OCP [V]") or pe_ocv(y).entries.mean()
+        self.stack_edd["Stack average OCP [V]"] = self.stack_edd.get("Stack average OCP [V]") or self.stack_edd.get("Positive electrode average OCP [V]") - self.stack_edd.get("Negative electrode average OCP [V]")
+
+        # areal capacity
+        ### add optional user defined 1st cycle capacity loss in case capacity is supplied explicitly or on top of "Formation capacity loss [mA.h.cm-2]"?
+        ### calculate capacity based on user defined cut-off voltages?
+        param = pybamm.LithiumIonParameters()
+        esoh_solver = pybamm.lithium_ion.ElectrodeSOHSolver(
+            self.parameter_values, param
+        )
+        Q_n = self.parameter_values.evaluate(param.n.Q_init)
+        Q_p = self.parameter_values.evaluate(param.p.Q_init)
+        Q_Li = self.parameter_values.evaluate(param.Q_Li_particles_init)
+        inputs_ = {"Q_Li": Q_Li, "Q_n": Q_n, "Q_p": Q_p}
+        sol = esoh_solver.solve(inputs_)
+        areal_capacity = sol[
+            "Capacity [mA.h.cm-2]" # or better min("sol[Q_n * (x_100 - x_0)]"", "sol[Q_p * (y_0 - y_100)]")?
+        ]
+        self.stack_edd["Capacity [mA.h.cm-2]"] = self.stack_edd.get("Capacity [mA.h.cm-2]") or areal_capacity
+
+        # thicknesses
+        if self.parameter_values.get("Negative current collector thickness [m]") is None:
```

Comment:
you can go `get(keyword, value` to return `value` if the keyword isn't found. My preference would not to be used hard coded defaults. this is where you should be using `inputs` then you can do something like 
```python3
if self.parameter_values.get("Negative current collector thickness [m]") is None:
    try:
        self.stack_edd["Negative current collector thickness [m]"] = inputs["Negative current collector thickness [m]"]
    except KeyError:
            raise ValueError("error message")
```